### PR TITLE
Standard Product 3

### PR DESCRIPTION
**Pull request by eggrobin**

### CODE REVIEW

File: astronomy/standard_product_3.hpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  struct SatelliteIdentifier {
+    SatelliteGroup group;
+    int index;
+
+    friend bool operator<(SatelliteIdentifier const& left,
+                          SatelliteIdentifier const& right) {
+      return left.group < right.group ||
+             (left.group == right.group && left.index < right.index);
+    }
```
Comment (pleroy) [inline astronomy/standard_product_3.hpp]: Put the implementation in the `.cc` file.

File: astronomy/standard_product_3.hpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  };
+
+  struct OrbitPoint {
```
Comment (pleroy) [inline astronomy/standard_product_3.hpp]: `OrbitDegreesOfFreedom`?  The word "point" has a specific technical meaning in our code base.

File: astronomy/standard_product_3.hpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    // Divergences from the specification:
+    // - the number of epochs given on the first line is off by one;
+    // - comments start with %/* instead of /*;
+    // - the fields of the epoch record are shifted one column to the left;
+    // - the minute field of the epoch record takes the value 60 at the end of
+    //   the hour.
+    ILRSB,
+  };
+
+  enum SatelliteGroup : char {
```
Comment (pleroy) [inline astronomy/standard_product_3.hpp]: I'd use an `enum class` here.  `SatelliteGroup::General` is going to work better than `StandardProduct3::General` or `General`.

File: astronomy/standard_product_3.hpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  };
+
+  StandardProduct3(std::filesystem::path const& filename, Dialect dialect);
+
+  std::vector<OrbitPoint> const& orbit(SatelliteIdentifier const& id) const;
+
+ private:
+  std::map<SatelliteIdentifier, std::vector<OrbitPoint>> orbits_;
+
+  // 'a' through 'd'.
```
Comment (pleroy) [inline astronomy/standard_product_3.hpp]: Could we have an enum for this?

File: astronomy/standard_product_3.cpp [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+using quantities::si::Kilo;
+using quantities::si::Metre;
+using quantities::si::Second;
+
+StandardProduct3::StandardProduct3(
+    std::filesystem::path const& filename,
+    StandardProduct3::Dialect const dialect) {
+  std::ifstream file(filename);
+  CHECK(file.good()) << filename;
+  std::optional<std::string> line;
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: Blank line after.

File: astronomy/standard_product_3.cpp [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+    } else {
+      ++line_number;
+      location =
+          absl::StrCat(filename.string(), " line ", line_number, ": ", *line);
+    }
+  };
+
+  // The specification uses 1-based column indices, and column ranges with
+  // bounds included.
+  auto const column = [&line, &location, line_number](int index) -> char {
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: Should this be called `get_column`?  Function typically use a verb.  Same question for the other lambdas.

Also, `int const`.

File: astronomy/standard_product_3.cpp [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    CHECK(line.has_value()) << location;
+    CHECK_LT(index - 1, line->size()) << location;
+    return (*line)[index - 1];
+  };
+  auto const columns = [&line, &location](
+                            int first, int last) -> std::string_view {
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: `int const` (2x).  Also, let's omit the return type when it's not necessary.

Do we believe that Mac has `string_view`?

File: astronomy/standard_product_3.cpp [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+    CHECK(line.has_value()) << location;
+    CHECK_LT(last - 1, line->size()) << location;
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: Should there be a check that `first` is before `last`?

File: astronomy/standard_product_3.cpp [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+    return std::string_view(&(*line)[first - 1], last - first + 1);
+  };
+  auto const float_columns = [&columns, &location](int first,
+                                                   int last) -> double {
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: `int const` (2x)

File: astronomy/standard_product_3.cpp [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+    double result;
+    CHECK(absl::SimpleAtod(columns(first, last), &result))
+        << location << " columns " << first << "-" << last;
+    return result;
+  };
+  auto const integer_columns = [&columns, &location](int first,
+                                                     int last) -> int {
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: `int const`

File: astronomy/standard_product_3.cpp [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+              CHECK_GE(version_, 'c') << full_location;
+              break;
+            default:
+              LOG(FATAL) << "invalid satellite identifier " << id << ": "
+                         << full_location;
+          }
+        }
+        id.index = integer_columns(c + 1, c + 2);
+        CHECK_GT(id.index, 0) << full_location;
+        CHECK(orbits_.emplace(std::piecewise_construct,
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: Use `std::tie` and `std::ignore`.

File: astronomy/standard_product_3.cpp [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+    } else if (time_system == "TAI") {
+      parse_time = &ParseTAI;
+    } else if (time_system == "BDT") {
+      parse_time = &Parse北斗Time;
+    } else if (time_system == "GPS" || time_system == "GAL" ||
+               time_system == "TAI" || time_system == "IRN" ||
+               time_system == "QZS") {
+      parse_time = &ParseGPSTime;
+    } else {
+      LOG(FATAL) << "unexpected time system identifier " << time_system << ": "
```
Comment (pleroy) [inline astronomy/standard_product_3.cpp]: Messages should start with a capital letter (several places in this PR).

File: astronomy/standard_product_3.hpp (hunk 2 above)
Comment (eggrobin) [reply astronomy/standard_product_3.hpp]: Making it a DiscreteTrajectory.

File: astronomy/standard_product_3.cpp (hunk 6 above)
Comment (eggrobin) [reply astronomy/standard_product_3.cpp]: Functions generally aren't verbs in this codebase.

File: astronomy/standard_product_3.cpp (hunk 7 above)
Comment (eggrobin) [reply astronomy/standard_product_3.cpp]: Done.

> Do we believe that Mac has string_view?

 Sort of...

File: astronomy/standard_product_3.cpp (hunk 11 above)
Comment (eggrobin) [reply astronomy/standard_product_3.cpp]: Let’s try https://en.cppreference.com/w/cpp/language/structured_binding and see what breaks.

**Pull request merged**